File: SudokuGenerator.cs
// SudokuGenerator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace SudokuGenerator
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            var gen = new SudokuGenerator();
            if (opts.Generate != null)
            {
                var solution = gen.Generate(opts.Generate);
                if (opts.Print)
                {
                    Console.WriteLine($"Generated {opts.Generate} puzzle:");
                    gen.Print();
                }
                if (opts.ExportJson != null)
                {
                    var data = new { puzzle = gen.Grid, solution };
                    string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
                    File.WriteAllText(opts.ExportJson, json);
                    Console.WriteLine($"Exported to {opts.ExportJson}");
                }
                if (opts.ExportCsv != null) gen.ExportCsv(opts.ExportCsv);
                if (opts.ExportTxt != null) gen.ExportTxt(opts.ExportTxt);
            }
            else if (opts.Solve != null)
            {
                gen.Load(opts.Solve);
                if (gen.Solve())
                {
                    if (opts.Print)
                    {
                        Console.WriteLine("Solved puzzle:");
                        gen.Print();
                    }
                    if (opts.ExportJson != null)
                    {
                        gen.ExportJson(opts.ExportJson);
                        Console.WriteLine($"Exported to {opts.ExportJson}");
                    }
                }
                else
                {
                    Console.WriteLine("No solution found.");
                }
            }
            else if (opts.Import != null)
            {
                gen.Load(opts.Import);
                if (opts.Print) gen.Print();
            }
            else
            {
                Console.WriteLine("Use --help for usage.");
            }
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--generate": opts.Generate = args[++i]; break;
                    case "--solve": opts.Solve = args[++i]; break;
                    case "--export-json": opts.ExportJson = args[++i]; break;
                    case "--export-csv": opts.ExportCsv = args[++i]; break;
                    case "--export-txt": opts.ExportTxt = args[++i]; break;
                    case "--import": opts.Import = args[++i]; break;
                    case "--print": opts.Print = true; break;
                }
            }
            return opts;
        }

        class Options
        {
            public string Generate { get; set; }
            public string Solve { get; set; }
            public string ExportJson { get; set; }
            public string ExportCsv { get; set; }
            public string ExportTxt { get; set; }
            public string Import { get; set; }
            public bool Print { get; set; }
        }

        class SudokuGenerator
        {
            private int[,] grid = new int[9, 9];
            private Random rand = new Random();
            public int[,] Grid => grid;

            public bool IsValid(int row, int col, int num)
            {
                for (int i = 0; i < 9; i++)
                    if (grid[row, i] == num || grid[i, col] == num) return false;
                int startRow = (row / 3) * 3;
                int startCol = (col / 3) * 3;
                for (int i = 0; i < 3; i++)
                    for (int j = 0; j < 3; j++)
                        if (grid[startRow + i, startCol + j] == num) return false;
                return true;
            }

            public bool Solve()
            {
                for (int row = 0; row < 9; row++)
                {
                    for (int col = 0; col < 9; col++)
                    {
                        if (grid[row, col] == 0)
                        {
                            for (int num = 1; num <= 9; num++)
                            {
                                if (IsValid(row, col, num))
                                {
                                    grid[row, col] = num;
                                    if (Solve()) return true;
                                    grid[row, col] = 0;
                                }
                            }
                            return false;
                        }
                    }
                }
                return true;
            }

            public int CountSolutions(int limit)
            {
                int count = 0;
                for (int row = 0; row < 9; row++)
                {
                    for (int col = 0; col < 9; col++)
                    {
                        if (grid[row, col] == 0)
                        {
                            for (int num = 1; num <= 9; num++)
                            {
                                if (IsValid(row, col, num))
                                {
                                    grid[row, col] = num;
                                    count += CountSolutions(limit);
                                    grid[row, col] = 0;
                                    if (count >= limit) return count;
                                }
                            }
                            return count;
                        }
                    }
                }
                return 1;
            }

            public int[,] Generate(string difficulty)
            {
                FillDiagonalBlocks();
                Solve();
                int[,] solution = (int[,])grid.Clone();
                int cellsToRemove = difficulty switch
                {
                    "easy" => 30,
                    "medium" => 40,
                    "hard" => 50,
                    "expert" => 55,
                    _ => 40
                };
                var positions = new List<(int, int)>();
                for (int r = 0; r < 9; r++)
                    for (int c = 0; c < 9; c++)
                        positions.Add((r, c));
                positions = positions.OrderBy(_ => rand.Next()).ToList();
                int removed = 0;
                foreach (var (r, c) in positions)
                {
                    if (removed >= cellsToRemove) break;
                    int backup = grid[r, c];
                    grid[r, c] = 0;
                    var copy = Copy();
                    if (copy.CountSolutions(2) == 1)
                    {
                        removed++;
                    }
                    else
                    {
                        grid[r, c] = backup;
                    }
                }
                return solution;
            }

            private SudokuGenerator Copy()
            {
                var copy = new SudokuGenerator();
                for (int i = 0; i < 9; i++)
                    for (int j = 0; j < 9; j++)
                        copy.grid[i, j] = grid[i, j];
                return copy;
            }

            private void FillDiagonalBlocks()
            {
                for (int block = 0; block < 9; block += 3)
                {
                    var nums = Enumerable.Range(1, 9).ToList();
                    nums = nums.OrderBy(_ => rand.Next()).ToList();
                    int idx = 0;
                    for (int i = block; i < block + 3; i++)
                        for (int j = block; j < block + 3; j++)
                            grid[i, j] = nums[idx++];
                }
            }

            public void Print(bool color = true)
            {
                string reset = color ? "\u001B[0m" : "";
                string blue = color ? "\u001B[34m" : "";
                string white = color ? "\u001B[37m" : "";
                for (int i = 0; i < 9; i++)
                {
                    string line = "";
                    for (int j = 0; j < 9; j++)
                    {
                        int val = grid[i, j];
                        string ch = val == 0 ? "." : val.ToString();
                        line += (val == 0 ? blue : white) + ch + reset + " ";
                        if (j == 2 || j == 5) line += "| ";
                    }
                    Console.WriteLine(line);
                    if (i == 2 || i == 5) Console.WriteLine("------+-------+------");
                }
            }

            public void ExportJson(string filename)
            {
                string json = JsonSerializer.Serialize(grid, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(filename, json);
            }

            public void ExportCsv(string filename)
            {
                using var sw = new StreamWriter(filename);
                for (int i = 0; i < 9; i++)
                {
                    for (int j = 0; j < 9; j++)
                    {
                        if (j > 0) sw.Write(",");
                        sw.Write(grid[i, j]);
                    }
                    sw.WriteLine();
                }
            }

            public void ExportTxt(string filename)
            {
                using var sw = new StreamWriter(filename);
                for (int i = 0; i < 9; i++)
                {
                    for (int j = 0; j < 9; j++)
                    {
                        if (j > 0) sw.Write(" ");
                        sw.Write(grid[i, j]);
                    }
                    sw.WriteLine();
                }
            }

            public void Load(string filename)
            {
                string content = File.ReadAllText(filename);
                if (filename.EndsWith(".json"))
                {
                    var data = JsonSerializer.Deserialize<int[][]>(content);
                    for (int i = 0; i < 9 && i < data.Length; i++)
                        for (int j = 0; j < 9 && j < data[i].Length; j++)
                            grid[i, j] = data[i][j];
                }
                else
                {
                    var lines = content.Split('\n');
                    for (int i = 0; i < 9 && i < lines.Length; i++)
                    {
                        var parts = lines[i].Trim().Split(new[] { ' ', ',' }, StringSplitOptions.RemoveEmptyEntries);
                        for (int j = 0; j < 9 && j < parts.Length; j++)
                            grid[i, j] = int.Parse(parts[j]);
                    }
                }
            }
        }
    }
}

File: sudoku_generator.rs
// sudoku_generator.rs
use clap::{App, Arg};
use rand::prelude::*;
use serde::{Deserialize, Serialize};
use serde_json;
use std::fs;
use std::io::Write;
use colored::*;

#[derive(Serialize, Deserialize, Clone)]
struct Sudoku {
    grid: [[u8; 9]; 9],
}

impl Sudoku {
    fn new() -> Self {
        Sudoku { grid: [[0; 9]; 9] }
    }

    fn copy(&self) -> Self {
        let mut g = [[0; 9]; 9];
        for i in 0..9 {
            for j in 0..9 {
                g[i][j] = self.grid[i][j];
            }
        }
        Sudoku { grid: g }
    }

    fn is_valid(&self, row: usize, col: usize, num: u8) -> bool {
        for i in 0..9 {
            if self.grid[row][i] == num || self.grid[i][col] == num {
                return false;
            }
        }
        let start_row = (row / 3) * 3;
        let start_col = (col / 3) * 3;
        for i in 0..3 {
            for j in 0..3 {
                if self.grid[start_row + i][start_col + j] == num {
                    return false;
                }
            }
        }
        true
    }

    fn solve(&mut self) -> bool {
        for row in 0..9 {
            for col in 0..9 {
                if self.grid[row][col] == 0 {
                    for num in 1..=9 {
                        if self.is_valid(row, col, num) {
                            self.grid[row][col] = num;
                            if self.solve() {
                                return true;
                            }
                            self.grid[row][col] = 0;
                        }
                    }
                    return false;
                }
            }
        }
        true
    }

    fn count_solutions(&mut self, limit: u8) -> u8 {
        let mut count = 0;
        for row in 0..9 {
            for col in 0..9 {
                if self.grid[row][col] == 0 {
                    for num in 1..=9 {
                        if self.is_valid(row, col, num) {
                            self.grid[row][col] = num;
                            count += self.count_solutions(limit);
                            self.grid[row][col] = 0;
                            if count >= limit {
                                return count;
                            }
                        }
                    }
                    return count;
                }
            }
        }
        1
    }

    fn generate(&mut self, difficulty: &str) -> [[u8; 9]; 9] {
        self.fill_diagonal_blocks();
        self.solve();
        let solution = self.grid;
        let cells_to_remove = match difficulty {
            "easy" => 30,
            "medium" => 40,
            "hard" => 50,
            "expert" => 55,
            _ => 40,
        };
        let mut positions = Vec::new();
        for r in 0..9 {
            for c in 0..9 {
                positions.push((r, c));
            }
        }
        positions.shuffle(&mut rand::thread_rng());
        let mut removed = 0;
        for (r, c) in positions {
            if removed >= cells_to_remove {
                break;
            }
            let backup = self.grid[r][c];
            self.grid[r][c] = 0;
            let mut copy = self.copy();
            if copy.count_solutions(2) == 1 {
                removed += 1;
            } else {
                self.grid[r][c] = backup;
            }
        }
        solution
    }

    fn fill_diagonal_blocks(&mut self) {
        let mut rng = rand::thread_rng();
        for block in (0..9).step_by(3) {
            let mut nums: Vec<u8> = (1..=9).collect();
            nums.shuffle(&mut rng);
            let mut idx = 0;
            for i in block..block + 3 {
                for j in block..block + 3 {
                    self.grid[i][j] = nums[idx];
                    idx += 1;
                }
            }
        }
    }

    fn print(&self, color: bool) {
        let reset = if color { "\x1b[0m" } else { "" };
        let blue = if color { "\x1b[34m" } else { "" };
        let white = if color { "\x1b[37m" } else { "" };
        for i in 0..9 {
            let mut line = String::new();
            for j in 0..9 {
                let val = self.grid[i][j];
                let ch = if val == 0 { ".".to_string() } else { val.to_string() };
                let colored = if val == 0 { blue } else { white };
                line.push_str(&format!("{}{}{} ", colored, ch, reset));
                if j == 2 || j == 5 {
                    line.push_str("| ");
                }
            }
            println!("{}", line);
            if i == 2 || i == 5 {
                println!("------+-------+------");
            }
        }
    }

    fn export_json(&self, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
        let json = serde_json::to_string_pretty(&self.grid)?;
        fs::write(filename, json)?;
        Ok(())
    }

    fn export_csv(&self, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut wtr = csv::Writer::from_path(filename)?;
        for row in &self.grid {
            let row_str: Vec<String> = row.iter().map(|v| v.to_string()).collect();
            wtr.write_record(&row_str)?;
        }
        wtr.flush()?;
        Ok(())
    }

    fn export_txt(&self, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut content = String::new();
        for row in &self.grid {
            for (i, v) in row.iter().enumerate() {
                if i > 0 {
                    content.push(' ');
                }
                content.push_str(&v.to_string());
            }
            content.push('\n');
        }
        fs::write(filename, content)?;
        Ok(())
    }

    fn load(filename: &str) -> Result<Self, Box<dyn std::error::Error>> {
        let content = fs::read_to_string(filename)?;
        let mut grid = [[0; 9]; 9];
        if filename.ends_with(".json") {
            let data: [[u8; 9]; 9] = serde_json::from_str(&content)?;
            grid = data;
        } else {
            for (i, line) in content.lines().enumerate() {
                if i >= 9 { break; }
                let nums: Vec<u8> = line.split_whitespace()
                    .filter_map(|s| s.parse().ok())
                    .collect();
                for (j, &v) in nums.iter().enumerate() {
                    if j < 9 { grid[i][j] = v; }
                }
            }
        }
        Ok(Sudoku { grid })
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matches = App::new("Sudoku Generator")
        .arg(Arg::with_name("generate").long("generate").takes_value(true).possible_values(&["easy", "medium", "hard", "expert"]))
        .arg(Arg::with_name("solve").long("solve").takes_value(true))
        .arg(Arg::with_name("export-json").long("export-json").takes_value(true))
        .arg(Arg::with_name("export-csv").long("export-csv").takes_value(true))
        .arg(Arg::with_name("export-txt").long("export-txt").takes_value(true))
        .arg(Arg::with_name("import").long("import").takes_value(true))
        .arg(Arg::with_name("print").long("print"))
        .get_matches();

    let mut sudoku = Sudoku::new();

    if let Some(level) = matches.value_of("generate") {
        let solution = sudoku.generate(level);
        if matches.is_present("print") {
            println!("Generated {} puzzle:", level);
            sudoku.print(true);
        }
        if let Some(file) = matches.value_of("export-json") {
            let data = serde_json::json!({ "puzzle": sudoku.grid, "solution": solution });
            fs::write(file, serde_json::to_string_pretty(&data)?)?;
            println!("Exported to {}", file);
        }
        if let Some(file) = matches.value_of("export-csv") {
            sudoku.export_csv(file)?;
        }
        if let Some(file) = matches.value_of("export-txt") {
            sudoku.export_txt(file)?;
        }
    } else if let Some(file) = matches.value_of("solve") {
        let mut s = Sudoku::load(file)?;
        if s.solve() {
            if matches.is_present("print") {
                println!("Solved puzzle:");
                s.print(true);
            }
            if let Some(out) = matches.value_of("export-json") {
                s.export_json(out)?;
                println!("Exported to {}", out);
            }
        } else {
            println!("No solution found.");
        }
    } else if let Some(file) = matches.value_of("import") {
        let s = Sudoku::load(file)?;
        sudoku = s;
        if matches.is_present("print") {
            sudoku.print(true);
        }
    } else {
        println!("Use --help for usage.");
    }
    Ok(())
}
